packages/flutter_tools/lib/src/android/android_sdk.dart

```diff
@@ -104,6 +104,23 @@ String getAvdPath() {
   );
 }
 
+/// Locate 'avdmanager'. Prefer to use one from an Android SDK, if we can locate that.
+/// This should be used over accessing androidSdk.avdManagerPath directly because it
+/// will work for those users who have Android Tools installed but
+/// not the full SDK.
+String getAvdManagerPath([AndroidSdk existingSdk]) {
+  if (existingSdk?.avdManagerPath != null)
+    return existingSdk.avdManagerPath;
+
+  final AndroidSdk sdk = AndroidSdk.locateAndroidSdk();
+
+  if (sdk?.latestVersion == null) {
```

Comment:
I suspect this fallback logic parallels other android tools lookups in the source base (adb?). We may want to just fail if we can't locate an android sdk however (not do the path based lookup for `avdmanager`).

Reply:
Currently there are three possibilities in here (same for the other few in this file):

1. If it exists on the SDK passed in (which looks to come from `context[AndroidSdk]`), we use that
2. If we can locate an SDK, that's used
3. Otherwise we try `os.which('emulator')?.path`

I'm not sure which out 1 and 2 is the best one (or whether we need 3). WDYT? If going with 1 then we can just delete this function; but I'm not sure whether that's a good bet. There are some other paths that do seem to use that and bail out if it's not set without any fallback, eg.:

```dart
  Future<bool> _checkForSupportedAdbVersion() async {
    if (androidSdk == null)
      return false;
```

Though interestingly, it then calls `getAdbPath` which does the fallback; but in this case it would never trigger because of the `return` above.

Reply:
Let's just delete the `os.which('emulator')?.path` option.

> Though interestingly, it then calls getAdbPath which does the fallback; but in this case it would never trigger because of the return above.

Hmm, sounds like some logic was orphaned by a refactor. Falling back to 'which `adb`' was useful since the user could then have a little functionality - like listing devices - even if they didn't have the full functionality that configuring an android sdk would give. If this regressed at some point, and people didn't notice, than it's likely that we don't need the fallback even for `adb`.

Reply:
Ok, I'll remove the `which`. Looks like there are still some code paths that don't do what I described above, including `getAdbDevices`, so my comment about was a little misleading. Whether or not `_checkForSupportedAdbVersion` is behaving as we'd like though, I'm not sure. LMK if you think this needs some investigation and I'll take a look outside of this PR.